packages/kokkos/core/src/impl/Kokkos_stacktrace.cpp

```diff
@@ -0,0 +1,265 @@
+  ~Stacktrace () = delete;
+
+  // These are public only to avoid wasting an extra stacktrace line.
+  // See save_stacktrace below.
+  static constexpr int capacity = 100;
+  static void* buffer[capacity];
+  static int length;
+
+  static std::vector<std::string> lines ()
+  {
+    char** symbols = backtrace_symbols (buffer, length);
+    if (symbols == nullptr) {
+      return {};
+    }
+    else {
+      std::vector<std::string> trace (length);
+      for (int i = 0; i < length; ++i) {
+        if (symbols[i] != nullptr) {
+          trace[i] = std::string (symbols[i]);
+        }
+      }
+      free (symbols);
+      return trace;
+    }
+  }
+};
+
+int Stacktrace::length = 0;
+void* Stacktrace::buffer[Stacktrace::capacity];
+
+void save_stacktrace ()
+{
+  Stacktrace::length = backtrace (Stacktrace::buffer,
+                                  Stacktrace::capacity);
+}
+
+size_t
+find_first_non_whitespace (const std::string& s, const size_t start_pos)
+{
+  constexpr size_t num_ws_chars = 3;
+  const char ws_chars[] = "\n\t ";
+  return s.find_first_not_of (ws_chars, start_pos, num_ws_chars);
+}
+
+size_t
+find_first_whitespace (const std::string& s, const size_t start_pos)
+{
+  constexpr size_t num_ws_chars = 3;
+  const char ws_chars[] = "\n\t ";
+  return s.find_first_of (ws_chars, start_pos, num_ws_chars);
+}
+
+template<class Callback>
+void for_each_token (const std::string& s, Callback c) {
+  size_t cur = find_first_non_whitespace (s, 0);
+  while (cur != std::string::npos) {
+    const size_t end = find_first_whitespace (s, cur);
+    const bool last = (end == std::string::npos);
+    const size_t count = last ? end : size_t (end - cur);
+    c (s.substr (cur, count), last);
+    cur = find_first_non_whitespace (s, end);
+  }
+}
+
+// Search the whole backtrace, column by column, for "main".
+// This tells us what column has the function names.
+// While we're doing that, figure out the longest column,
+// so we can compute spacing correctly.
+
+std::tuple<bool, size_t, std::vector<size_t>>
+find_main_column (const std::vector<std::string>& traceback)
+{
+  bool found_main = false;
+  size_t main_col = 0;
+  for (auto&& entry : traceback) {
+    size_t col_count = 0;
+    for_each_token (entry, [&] (const std::string& s, bool) {
+                             const size_t pos = s.find ("main");
+                             if (pos != std::string::npos) {
+                               found_main = true;
+                               main_col = col_count;
+                             }
+                             ++col_count;
+                           });
+    if (found_main) {
+      break;
+    }
+  }
+
+  // Make another pass to get the column lengths.
+  // Only demangle the column of functions.
+  std::vector<size_t> max_col_lens;
+  for (auto&& entry : traceback) {
+    size_t col_count = 0;
+    for_each_token (entry, [&] (const std::string& s, bool) {
+                             const size_t cur_col_len =
```

Comment:
Wow, that is some crazy indentation.  What so much?

Reply:
Emacs wanted it that way, I guess?  I can fix that.

Reply:
Yes, default email tab indentation is insane.  I wrote a much more sane indentation scheme for emails.  You can find it at:

https://github.com/trilinos/Trilinos/tree/master/packages/thyra/emacs

Should be a 5-min setup.